pkg/pipeline/stream/destination.go

```diff
@@ -51,23 +50,30 @@ func (n *DestinationNode) Run(ctx context.Context) (err error) {
 	if err != nil {
 		return cerrors.Errorf("could not open destination connector: %w", err)
 	}
-	defer func() {
-		// wait for acker node to receive all outstanding acks, time out after
-		// 1 minute or right away if the context is already canceled.
-		waitCtx, cancel := context.WithTimeout(ctx, time.Minute)
-		defer cancel()
-		n.AckerNode.Wait(waitCtx)
 
-		// teardown will kill the plugin process
-		tdErr := n.Destination.Teardown(connectorCtx)
-		if tdErr != nil {
+	// lastPosition stores the position of the last successfully processed record
+	var lastPosition record.Position
+	// openMsgTracker tracks open messages until they are acked or nacked
+	var openMsgTracker OpenMessagesTracker
+	defer func() {
+		stopErr := n.Destination.Stop(connectorCtx, lastPosition)
+		if stopErr != nil {
+			// log this error right away because we're not sure the connector
+			// will be able to stop right away, we might block for 1 minute
+			// waiting for acks and we don't want the log to be empty
+			n.logger.Err(ctx, err).Msg("could not stop destination connector")
 			if err == nil {
-				err = tdErr
-			} else {
-				// we are already returning an error, just log this error
-				n.logger.Err(ctx, err).Msg("could not tear down destination connector")
+				err = stopErr
 			}
 		}
+
+		openMsgTracker.Wait()
```

Comment:
First, the deferred function will be called regardless of the error that's returned (in other words - it's called _always_ when the node stops). Now, is it possible that it waits indefinitely? I think this would be possible, yes, if the connector just starts hanging (i.e. it's still running, looks alive, not really broken, but just hangs). On the other hand, if the connector stops because of an error or panic, then `DestinationAckerNode` will [nack all messages](https://github.com/ConduitIO/conduit/blob/8ad3b02a9cfbdb8fa3549b6cb4c88d361dcb111e/pkg/pipeline/stream/destination_acker.go#L178) that didn't get acked/nacked by the connector, thus unblocking this line.

If we wanted to fix the possibility of the pipeline getting stuck we would need to provide an option to "force stop" the pipeline (i.e. we wouldn't care what the connector is doing and just kill it). This would work for standalone plugins since we can kill the subprocess, while there's no way for us to kill built-in plugins (not that I know of at least), best we could do is to stop the nodes and "detach" from the plugin goroutines (leave them running in the background).

Reply:
> If we wanted to fix the possibility of the pipeline getting stuck

IMHO, that would be a good thing to do, not necessarily in this PR though.

Reply:
Yeah we should do that, though this is probably a bit lower on the priority list.